### PR TITLE
Lena hw4

### PR DESCRIPTION
**Pull request by arc-lena**
done task4

### CODE REVIEW

File: lena/homework_4/HW4/HW4/items.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Define here the models for your scraped items
+#
+
+import scrapy
+from scrapy import Item, Field
+
+
+class tem(scrapy.Item):
```
Comment (DmitryPoliuha) [inline lena/homework_4/HW4/HW4/items.py:8]: not the best naming here

File: lena/homework_4/HW4/HW4/spiders/crawler.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import scrapy
+from .. items import tem
```
Comment (DmitryPoliuha) [inline lena/homework_4/HW4/HW4/spiders/crawler.py:2]: What does `tem` stands for?

File: lena/homework_4/HW4/HW4/spiders/crawler.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    custom_settings = {
+        'FEED_URI': 'output.json',
+        'FEED_EXPORT_ENCODING': 'utf-8',
+        'CLOSESPIDER_ITEMCOUNT': 100,
+    }
+
+    count = 0
+
+    def parse(self, response):
+        l = []
```
Comment (DmitryPoliuha) [inline lena/homework_4/HW4/HW4/spiders/crawler.py:20]: `l`?

File: lena/homework_4/HW4/HW4/spiders/crawler.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        for link in response.css('div h3 a::attr(href)').extract():
+            url = response.urljoin(link)
+            l.append(url)
+        for link_prod in l:
+            yield scrapy.Request(link_prod, callback=self.parse0)
```
Comment (DmitryPoliuha) [inline lena/homework_4/HW4/HW4/spiders/crawler.py:25]: do you really need two loops here?

File: lena/homework_4/HW4/HW4/spiders/crawler.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        pagination_links = response.css('span.fleft a')
+        yield from response.follow_all(pagination_links, self.parse)
+
+    def parse0(self, response):
```
Comment (DmitryPoliuha) [inline lena/homework_4/HW4/HW4/spiders/crawler.py:29]: `parse0`?

File: lena/homework_4/HW4/HW4/spiders/crawler.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        if self.count<100:
+            for info in response.css("div.offerdescription"):
```
Comment (DmitryPoliuha) [inline lena/homework_4/HW4/HW4/spiders/crawler.py:31]: why do you need `for` loop here?

File: lena/homework_4/HW4/HW4/spiders/crawler.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+                product['olx_delivery_availability'] = olx_delivery_availability
+                product['urls_product'] = urls_product
+                product['seller_name'] = seller_name
+                product['seller_address'] = seller_address
+
+
+                self.count += 1
+                print(self.count)
+                yield product
+                if self.count == 100:
```
Comment (DmitryPoliuha) [inline lena/homework_4/HW4/HW4/spiders/crawler.py:64]: You've already checked that on line 30. Why do you need that here?

File: lena/homework_4/HW4/test.txt [hunk 8]
```diff
@@ -0,0 +1,10 @@
+{"product_price": "2 200 грн.", "product_name": "Левый наушник airpods pro, комплект 5580, (MWP22), идеальный звук ", "product_category": "Навушники", "product_photos": "https://static.olx.ua/static/olxua/naspersclassifieds-regional/olxeu-atlas-web/static/svg/redesign/delivery/delivery-badge.svg", "product_description": "Оригинал. Любые проверки . Комплект утерян , остался один наушник . Состояние идеальное , аккумулятор в полном порядке , заряжается на 100 % Следов эксплуатации нет . Пользовался 1 мес не часто . Только Киев , любая проверка при встречи . Или отправка по 100% предоплате .Или отправка олх доставкой", "product_id": "666000150", "produc_date_of_creation": "в 15:18, 9 жовтня 2020", "product_views_count": "324", "olx_delivery_availability": "", "urls_product": "https://www.olx.ua/obyavlenie/levyy-naushnik-airpods-pro-komplekt-5580-mwp22-idealnyy-zvuk-IDJ4t3U.html", "seller_name": "", "seller_address": null}
```
Comment (DmitryPoliuha) [inline lena/homework_4/HW4/test.txt:1]: As I remember we decided to have 2 fields for price

Comment (DmitryPoliuha) [reply lena/homework_4/HW4/test.txt:1]: `product_photos` should contain an array of urls, not 1 url

Comment (DmitryPoliuha) [reply lena/homework_4/HW4/test.txt:1]: `olx_delivery_availability` is blank for each product

Comment (DmitryPoliuha) [reply lena/homework_4/HW4/test.txt:1]: `seller_name` is blank for each product

Comment (DmitryPoliuha) [reply lena/homework_4/HW4/test.txt:1]: `product_views_count` should be int

**Review by DimaPoliuha — COMMENTED** [review]
You need to remove all `.pyc` files from PR. You used weird naming for methods/variables etc. Not the best name of the project `Hw4`. OLX spider might be better. 
Your crawler has some issues

**Review by DimaPoliuha — APPROVED** [review]
Nice. It would be better to have only one setting: settings.py or custom_settings here. 
One more thing - could you check the output of `product_category` in your results?

**Review by PolyMaG — APPROVED** [review]

File: lena/homework_4/OLX_spider/OLX_spider/items.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
+# Define here the models for your scraped items
+#
+
+import scrapy
+from scrapy import Item, Field
```
Comment (andrey-ned) [inline lena/homework_4/OLX_spider/OLX_spider/items.py:5]: can be 1 import

File: lena/homework_4/OLX_spider/OLX_spider/items.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class CategoryItem(scrapy.Item):
```
Comment (andrey-ned) [inline lena/homework_4/OLX_spider/OLX_spider/items.py:8]: You impoty Item and use scrapy.Item

File: lena/homework_4/OLX_spider/OLX_spider/output.txt [hunk 11]
```diff
@@ -0,0 +1,20 @@
+{"product_price": 295, "product_currency": "грн.", "product_name": "MiraScreen Miracast HDMI-адаптер WiFi для подключения телефона к ТВ TV ", "product_category": "Бізнес", "product_photos": ["https://ireland.apollo.olxcdn.com:443/v1/files/ms9p0xb8wmjm1-UA/image;s=1000x700", "https://ireland.apollo.olxcdn.com:443/v1/files/jvbxk8r7va5y-UA/image;s=1000x700", "https://ireland.apollo.olxcdn.com:443/v1/files/7wxvtppj6lzx1-UA/image;s=1000x700", "https://ireland.apollo.olxcdn.com:443/v1/files/208b0x3g5k3e2-UA/image;s=1000x700"], "product_description": "MiraScreen G2 Wireless HDMI Dongle Miracast Airplay DLNA", "product_id": "555090510", "product_date_of_creation": "в 16:53, 10 жовтня 2020", "product_views_count": 8128, "olx_delivery_availability": "", "urls_product": "https://www.olx.ua/obyavlenie/mirascreen-miracast-hdmi-adapter-wifi-dlya-podklyucheniya-telefona-k-tv-tv-IDBz6oK.html", "seller_name": "Максим", "seller_address": "Одеса, Одеська область, Приморський"}
```
Comment (andrey-ned) [inline lena/homework_4/OLX_spider/OLX_spider/output.txt:1]: Price in Float - DIMA

File: lena/homework_4/OLX_spider/OLX_spider/spiders/olx_sp.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+from .. items import CategoryItem
+from scrapy.exceptions import CloseSpider
+
+class OLXSpider(scrapy.Spider):
+    name = "olx_el"
+    allowed_domains = ['olx.ua']
+    start_urls = [
+        'https://www.olx.ua/uk/elektronika/',
+    ]
+    custom_settings = {
```
Comment (andrey-ned) [inline lena/homework_4/OLX_spider/OLX_spider/spiders/olx_sp.py:11]: File settings for this

**Review by andrey-ned — COMMENTED** [review]
Where is your Req.txt?

**Pull request merged**